frontend/pages/policies/ManagePoliciesPage/components/PoliciesPaginatedList/PoliciesPaginatedList.tsx

```diff
@@ -0,0 +1,198 @@
+import { useQueryClient } from "react-query";
+import { IPolicy } from "interfaces/policy";
+import teamPoliciesAPI from "services/entities/team_policies";
+import globalPoliciesAPI from "services/entities/global_policies";
+
+import { APP_CONTEXT_ALL_TEAMS_ID } from "interfaces/team";
+import Button from "components/buttons/Button";
+import TooltipWrapper from "components/TooltipWrapper";
+
+// Extend the IPolicy interface with some virtual properties that make it easier
+// to track item state. These are set by the various Manage Automations modals.
+export interface IFormPolicy extends IPolicy {
+  installSoftwareEnabled: boolean;
+  swIdToInstall?: number;
+  runScriptEnabled: boolean;
+  scriptIdToRun?: number;
+}
+
+interface IPoliciesPaginatedListProps {
+  isSelected: string | ((item: IFormPolicy) => boolean);
+  renderItemRow?: (
+    item: IFormPolicy,
+    onChange: (item: IFormPolicy) => void
+  ) => ReactElement | false | null | undefined;
+  onToggleItem: (item: IFormPolicy) => IFormPolicy;
+  onCancel: () => void;
+  onSubmit: (formData: IFormPolicy[]) => void;
+  isUpdating: boolean;
+  disableSave?: (changedItems: IFormPolicy[]) => boolean | string;
+  teamId: number;
+  footer: ReactElement | undefined | null;
+}
+
+const baseClass = "policies-paginated-list";
+
+function PoliciesPaginatedList(
+  {
+    isSelected,
+    renderItemRow,
+    onToggleItem,
+    onCancel,
+    onSubmit,
+    isUpdating,
+    disableSave,
+    teamId,
+    footer,
+  }: IPoliciesPaginatedListProps,
+  ref: Ref<IPaginatedListHandle<IFormPolicy>>
+) {
+  // Create a ref to use with the PaginatedList, so we can access its dirty items.
+  const paginatedListRef = useRef<IPaginatedListHandle<IFormPolicy>>(null);
+
+  const [saveDisabled, setSaveDisabled] = useState<string | boolean>(false);
+
+  // Allow parents access to the `getDirtyItems` of the underlying PaginatedList.
+  useImperativeHandle(ref, () => ({
+    getDirtyItems() {
+      if (paginatedListRef.current) {
+        return paginatedListRef.current.getDirtyItems();
+      }
+      return [];
+    },
+  }));
+
+  // When "save" is clicked, call the parent's `onSubmit` with the set of changed items.
+  const onClickSave = () => {
+    let changedItems: IFormPolicy[] = [];
+    if (paginatedListRef.current) {
+      changedItems = paginatedListRef.current.getDirtyItems();
+    }
+    onSubmit(changedItems);
+  };
+
+  // When any item in the list is changed, check whether we should disable the save button.
+  const onUpdate = useCallback(
+    (changedItems: IFormPolicy[]) => {
+      if (!disableSave) {
+        return;
+      }
+      setSaveDisabled(disableSave(changedItems));
+    },
+    [disableSave]
+  );
+
+  // Fetch a single page of policies.
+  const queryClient = useQueryClient();
+  const DEFAULT_PAGE_SIZE = 8;
+  const DEFAULT_SORT_COLUMN = "name";
+
+  const fetchPage = useCallback((pageNumber: number) => {
+    let fetchPromise;
+
+    if (teamId === APP_CONTEXT_ALL_TEAMS_ID) {
```

Comment:
Very cool to see how we can interact a bit lower down with `useQueryClient`! Have you checked out how we [conditionally fetch global or team queries](https://github.com/fleetdm/fleet/blob/b990b3c6d92092854a709060b74a16f14b5133cd/frontend/pages/queries/ManageQueriesPage/ManageQueriesPage.tsx#L123-L155)? By setting the query key's `teamId` and `mergeInherited` fields based on `teamIdForApi`, we get the desired API calls. I think this would be a bit cleaner here, unless there's some specific reason to use `useQueryClient`?

Reply:
The queries API handles team vs. global queries on the back end, but alas the policies API has separate endpoints for team vs. global.  As for using `useQueryClient()`, if you mean as opposed to `useQuery()`, this is so that we can return a promise to the lower-level `PaginatedList` component, rather than doing the query automatically here in the PoliciesPaginatedList component.  The idea of `PaginatedList` is for it to contain all of the data and state itself, rather than the parent having to keep track of the data and what page we're on.